src/storage/src/source/mod.rs

```diff
@@ -801,72 +803,52 @@ where
                     // It's time to timestamp a batch
                     _ = timestamp_interval.tick() => {
-                        let mut max_offsets = HashMap::new();
-                        for message in pending_messages.iter().filter_map(|m| m.as_ref().ok()) {
-                            let entry = max_offsets.entry(message.partition.clone()).or_default();
-                            *entry = std::cmp::max(*entry, message.offset);
-                        }
-                        let (bindings, progress) = match timestamper.timestamp_offsets(&max_offsets).await {
-                            Ok((bindings, progress)) => (bindings, progress),
-                            Err(e) => {
-                                error!("Error timestamping offsets: {}", e);
-                                return;
-                            }
-                        };
-
-                        for msg in pending_messages.drain(..) {
-                            match msg{
-                                Ok(message) => {
-                                    let ts = bindings.get(&message.partition).expect("timestamper didn't return partition").0;
-                                    yield Event::Message(ts, Ok(message));
-                                },
-                                Err(e) => {
-                                    // TODO: make errors definite
-                                    yield Event::Message(0, Err(e));
-                                },
+                        println!("reclocking stuff");
+                        for (_, part_messages) in timestamper.reclock(&mut untimestamped_messages, &source_upper).await {
+                            for (message, ts) in part_messages {
+                                println!("producing messages");
+                                yield Event::Message(ts, Ok(message));
                             }
                         }
-                        let progress_some = progress.as_option().is_some();
-                        yield Event::Progress(progress.into_option());
-                        if source_stream.is_done() {
-                            // We just emitted the last piece of data that needed to be timestamped
-                            if progress_some {
+                        for err in non_definite_errors.drain(..) {
+                            // TODO: make errors definite
+                            yield Event::Message(0, Err(err));
+                        }
+                        // TODO(petrosag): compaction should be driven by AllowCompaction commands
+                        // coming from the storage controller
+                        let new_ts_upper = timestamper.advance(&source_upper).await;
+                        if PartialOrder::less_than(&ts_upper, &new_ts_upper) {
+                            ts_upper = new_ts_upper.clone();
+                            yield Event::Progress(new_ts_upper.into_option());
```

Comment:
```suggestion
                        // downgrade our capability to the new upper, but only if has changed since last time we downgraded it. Note property #2 documented above is critical here: once this capability is downgraded, that timestamp is sealed forever. 
                        if PartialOrder::less_than(&ts_upper, &new_ts_upper) {
                            ts_upper = new_ts_upper.clone();
                            yield Event::Progress(new_ts_upper.into_option());
```